langchain/vectorstores/pgvector.py

```diff
@@ -0,0 +1,421 @@
+
+    # custom_id : any user defined id
+    custom_id = sqlalchemy.Column(sqlalchemy.String, nullable=True)
+
+
+class QueryResult:
+    EmbeddingStore: EmbeddingStore
+    distance: float
+
+
+class DistanceStrategy(str, enum.Enum):
+    EUCLIDEAN = EmbeddingStore.embedding.l2_distance
+    COSINE = EmbeddingStore.embedding.cosine_distance
+    MAX_INNER_PRODUCT = EmbeddingStore.embedding.max_inner_product
+
+
+DEFAULT_DISTANCE_STRATEGY = DistanceStrategy.EUCLIDEAN
+
+
+@dataclass
+class PGVector(VectorStore):
+    """
+    VectorStore implementation using Postgres and pgvector.
+    - `connection_string` is a postgres connection string.
+    - `embedding_function` any embedding function implementing
+        `langchain.embeddings.base.Embeddings` interface.
+    - `collection_name` is the name of the collection to use. (default: langchain)
+        - NOTE: This is not the name of the table, but the name of the collection.
+            The tables will be created when initializing the store (if not exists)
+            So, make sure the user has the right permissions to create tables.
+    - `distance_strategy` is the distance strategy to use. (default: EUCLIDEAN)
+        - `EUCLIDEAN` is the euclidean distance.
+        - `COSINE` is the cosine distance.
+    - `pre_delete_collection` if True, will delete the collection if it exists.
+        (default: False)
+        - Useful for testing.
+    """
+
+    connection_string: str
+    embedding_function: Embeddings
+    collection_name: str = _LANGCHAIN_DEFAULT_COLLECTION_NAME
+    distance_strategy: DistanceStrategy = DEFAULT_DISTANCE_STRATEGY
+    pre_delete_collection: bool = False
+
+    def __post_init__(
+        self,
+    ) -> None:
+        """
+        Initialize the store.
+        """
+        self._conn = self.connect()
+        # self.create_vector_extension()
+        self.create_tables_if_not_exists()
+        self.create_collection()
+
+    def connect(self) -> sqlalchemy.engine.Connection:
+        engine = sqlalchemy.create_engine(self.connection_string)
+        conn = engine.connect()
+        return conn
+
+    def create_vector_extension(self) -> None:
+        try:
+            with Session(self._conn) as session:
+                statement = sqlalchemy.text("CREATE EXTENSION IF NOT EXISTS vector")
+                session.execute(statement)
+                session.commit()
+        except Exception as e:
+            print(e)
+
+    def create_tables_if_not_exists(self) -> None:
+        Base.metadata.create_all(self._conn)
+
+    def drop_tables(self) -> None:
+        Base.metadata.drop_all(self._conn)
+
+    def create_collection(self) -> None:
+        if self.pre_delete_collection:
+            self.delete_collection()
+        with Session(self._conn) as session:
+            CollectionStore.get_or_create(session, self.collection_name)
+
+    def delete_collection(self) -> None:
+        print("Trying to delete collection")
```

Comment:
better to use logger.info rather than print statements everywhere

Reply:
Yeah, will add that.